src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/RazorLSPHandlerBase.cs

```diff
@@ -0,0 +1,151 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.LanguageServer;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common;
+using Microsoft.VisualStudio.LanguageServer.Client;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Microsoft.VisualStudio.Threading;
+using OmniSharpPosition = OmniSharp.Extensions.LanguageServer.Protocol.Models.Position;
+
+namespace Microsoft.VisualStudio.LanguageServerClient.Razor.HtmlCSharp
+{
+    internal abstract class RazorLSPHandlerBase
+    {
+        private readonly int UndefinedDocumentVersion = -1;
+
+        public RazorLSPHandlerBase(
```

Comment:
I'd consider breaking this class into two services (I know it differs from our TypeScript impl), a service that invokes Tina's code (`LanguageServerInvoker` or something like that) and another that wraps our projection logic (`LSPProjectionProvider`). That way every LSP feature we add wont need to have a massive constructor that forwards additional information down into the base and it'll make unit testing the top-level constructs easier.

Reply:
Tried going this route but realized it wouldn't save us much if any because we'll need almost all of the types in each handler.

Reply:
I'm not super convinced but let's try this out for now to see how things play out

Reply:
This is what the constructor for any given handler we write would look like if we went that route,
```C#
        public AnotherHandler(
            JoinableTaskContext joinableTaskContext,
            ILanguageClientBroker languageClientBroker,
            LSPDocumentManager documentManager,
            LSPProjectionProvider projectionProvider,
            RazorLogger logger) { ... }
```

Reason is any given handler would most likely run on the main thread, get the projection, get the hostdocument, and execute RequestAsync and log something.

So we don't really save the constructor size. Am I missing something?

Reply:
I was thinking about it from a few aspects. Brain dump incoming

1. Currently your projection method takes a document snapshot, it could technically take a URI for the document and then do all the boiler plate internally. That'd enable your handler methods to look like:
```C#
public AnotherHandler(
    LSPRequestInvoker requestInvoker,
    LSPProjectionProvider projectionProvider,
    RazorLogger logger)
```

2. Complexity, getting projections in a correct way is one of the largest pieces of work we have to do for each feature.  Behind the scenes getting projections also depends on a complex piece of code which is asking an LSP middle layer to invoke requests on our behalf. Those two pieces of code are individually complex and will probably have a high level of complexity for each of their implementations. Having services for them ensures we isolate that complexity to make it easier to control / change in the future.

3. Separation, every one of our HTML/C# features are going to need to get a projection, reinvoke requests, most times both, sometimes only get a projection (language server middleware) etc. 

4. Testing, given the complexity of each piece, having quality testing types for each one can enable us to ensure each complex piece is working as expected. It could also enable us to integration test things we wouldn't normally integration test. For instance, imagine we build a `TestLSPRequestInvoker` that understands when it's asked to do a language query requests and could actually call through to our `RazorLanguageEndpoint`; for other requests it could be mocked appropriately for each feature.

All that said, this is just me putting my brain down on paper. I'm totally cool with trying out what you currently have and then only making adjustments in the future if needed after we've done a few features.

Reply:
Either way we definitely need to test the pieces in this class (`RequestServerAsync`, `GetProjectionAsync`) more directly since they lay the groundwork for all of our future LSP features.

Reply:
You convinced me especially with the testing argument. Updated.

Reply:
> Updated. I didn't build the `TestLSPRequestInvoker` though. It wouldn't be a unit test if we called into `RazorLanguageEndpoint`.

Super duper agree!